### PR TITLE
Add clojure LED support

### PR DESCRIPTION
**Pull request by Carrigan**
## Why?

- Because the Storefront installation has 540 LEDs that need to be controlled.
- Issue #24

## What Changed?

- Modified the Arduino firmware to drive all 540 LEDs and support for three commands for drawing on them.
- Added Clojure support for driving the LED panels over serial. The serial port can be designated using the `--serial <port here>` command line option.
- When LED panels are not present, they can be mocked using the same interface by adding the command line option `--mock`

---
Note that the firmware is not 100% polished and tested yet but I wanted to let people start using the LED mocks ASAP.

### CODE REVIEW

**Comment by sweetmandm** [comment]
I haven't fully grokked this but the interface into the C stuff looks good!

A couple tangentially related things i've been thinking about:
* has there been some thought about a way to draw shapes & antialiased lines? Like essentially an immediate mode graphics api that's rasterized by the c library. I haven't explored that idea much but you might be able to get away with just points, lines, polylines, fills, and transforms. or is the idea that it's better to rasterize on the clojure side and just pass the pixels?
* It'd be cool to store the previous LED frame so we can get some blurring effects & help smooth out the motion

**Comment by Carrigan** [comment]
- That should definitely be possible, although the gap in geometry would make the line functions kind of funky. Since the clojure->led panel interface is over serial, I would definitely prefer to do this in C and send the line's parameters rather than all of the pixels that the line touches.
- I love the idea of storing the previous LED frame, and that should be pretty trivial. We can definitely add this for the Durham edition.

File: arduino/strobe/led_strip.h [hunk 1]
```diff
@@ -3,9 +3,9 @@
 
 #include "color.h"
 
-// Default the LED_COUNT to 2m of 60 led/m
+// Set the LED count to 540, 18m @ 30/m
```
Comment (sweetmandm) [inline arduino/strobe/led_strip.h]: perhaps it's just me but it took me some time to realize the symbols following 540 meant 'strip length by led interval'. Maybe some more english would help, like `18 meter strip @ 30 led/meter`

File: arduino/strobe/strobe.ino [hunk 2]
```diff
@@ -1,62 +1,77 @@
...
-void setup() {
-  SPI.begin();
-  led_init(&strip, led_transfer);
+void clear_led_f(uint16_t index, color_t *color)
+{
+  memset(color, 0, sizeof(color_t));
 }
 
-int offset_count(int current, int led)
+void clear_leds(void *_none)
```
Comment (sweetmandm) [inline arduino/strobe/strobe.ino]: Why take `void *_none`? Is it to keep a consistent interface?

File: arduino/strobe/terminal.c [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  }
+
+  // Handle the incoming data
+  if(term->current_handler) {
+    term->buffer[term->character_index] = incoming;
+    term->character_index++;
+    if(term->character_index > term->current_handler->length) {
+      term->character_index = 0;
+      term->current_handler = NULL;
+      term->current_handler->handler((void *)term->buffer);
```
Comment (sweetmandm) [inline arduino/strobe/terminal.c]: ah, this looks like it answers why question about why the handlers need a consistent interface

**Comment by sweetmandm** [comment]
This c reads extremely well, it's like reading Chekhov :+1: I need to catch up on some clojure so you probably should get an extra pair of eyes on that

File: arduino/strobe/led_strip.h (hunk 1 above)
Comment (Carrigan) [inline arduino/strobe/led_strip.h]: :+1: I'll change the expression to `(18 * 30)` and English better

File: arduino/strobe/terminal.c (hunk 3 above)
Comment (Carrigan) [inline arduino/strobe/terminal.c]: :+1:

**Comment by Carrigan** [comment]
Thanks for the review @sweetmandm!

**Comment by dkniffin** [comment]
The clojure looks good, but the mocking doesn't seem to be working locally:

````
✘ |10015|  ~/Repositories/storefront ⎇ led-arrays > lein run cycle --mock
WARNING:  RXTX Version mismatch
	Jar version = RXTX-2.2pre2
	native lib Version = RXTX-2.2pre3
Unknown option: "--mock"
✘ |10016|  ~/Repositories/storefront ⎇ led-arrays > lein run textify --mock
WARNING:  RXTX Version mismatch
	Jar version = RXTX-2.2pre2
	native lib Version = RXTX-2.2pre3
Unknown option: "--mock"
✘ |10017|  ~/Repositories/storefront ⎇ led-arrays > lein run tweetreader --mock
WARNING:  RXTX Version mismatch
	Jar version = RXTX-2.2pre2
	native lib Version = RXTX-2.2pre3
Unknown option: "--mock"
````

**Comment by dkniffin** [comment]
Nevermind. It works. :shipit:

**Pull request merged**